### PR TITLE
adding nvidia models

### PR DESCRIPTION
**Pull request by nv-kkudrynski**

### CODE REVIEW

File: scripts/run_pytorch.sh [hunk 1]
```diff
@@ -6,12 +6,14 @@ conda activate base
...
+CUDAs='nvidia'
+
 
 for f in $ALL_FILE
 do
   echo "Running pytorch example in $f"
-  # FIXME: GAN models checkpoints are on cuda.
-  if [[ $f = $GANs* ]]; then
+  # FIXME: GAN and NVIDIA models checkpoints are on cuda.
+  if [[ $f = $GANs* ]] || [[ $f = $CUDAs* ]]; then
```
Comment (ailzhang) [inline scripts/run_pytorch.sh]: Adding a cuda checkpoint will require user have a cuda pytorch installed to load it. Would it possible to convert them to cpu in hubconf entrypoints and relax it here?

**Comment by ailzhang** [comment]
@nv-kkudrynski Also I'm not sure why this didn't trigger the circleCi check. See #23 as an example, maybe try rebasing?

File: nvidia_deeplearningexamples_tacotron2.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+author: NVIDIA
+tags: [audio]
+github-link: https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/SpeechSynthesis/Tacotron2
+featured_image_1: no-image
+featured_image_2: no-image
+---
+
+```python
+import torch
+hub_model = torch.hub.load(github='nvidia/DeepLearningExamples', model='nvidia_tacotron2')
```
Comment (ailzhang) [inline nvidia_deeplearningexamples_tacotron2.md]: These two args are positional args, you can remove the keywords I think

File: nvidia_deeplearningexamples_tacotron2.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
+---
+layout: hub_detail
+background-class: hub-background
+body-class: hub
+title: NCF
```
Comment (ailzhang) [inline nvidia_deeplearningexamples_tacotron2.md]: Title doesn't match model ;)

File: nvidia_deeplearningexamples_waveglow.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
+---
+layout: hub_detail
+background-class: hub-background
+body-class: hub
+title: NCF
```
Comment (ailzhang) [inline nvidia_deeplearningexamples_waveglow.md]: Title doesn't match model ;)

File: nvidia_deeplearningexamples_tacotron2.md (hunk 3 above)
Comment (nv-kkudrynski) [reply nvidia_deeplearningexamples_tacotron2.md]: sure. i already noticed it on netlify. corrected. thx

File: nvidia_deeplearningexamples_waveglow.md (hunk 4 above)
Comment (nv-kkudrynski) [reply nvidia_deeplearningexamples_waveglow.md]: sure. i already noticed it on netlify. corrected. thx

**Comment by gottbrath** [comment]
@soumith  -- are we going to be able to accept this whole pull request? I believe from side conversation you had some concerns with the NCF. Have those been addressed? 

If we can't accept the NCF part would it make sense to split this so that we could get the waveglow and tacotron2 models up first then address the NCF issues?

**Comment by soumith** [comment]
yes, there are a few changes needed.

1. can we remove NCF from this PR for now -- I dont think the model as it stands is useful for Hub (we are curating), because it isn't really useful for building research on top of it -- or useful to visualize anything about the output
2. Can you add code that visualizes the outputs of WaveGlow / TacoTron-2. They produce spectrograms, that's great -- it'd be useful to see that. You can use the GAN model that I cleaned up as an example: https://pytorch.org/hub/facebookresearch_pytorch-gan-zoo_dcgan/ . This way, people pulling your model know what to do with it further.

File: nvidia_deeplearningexamples_tacotron2.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+### Example
+
+Here's a sample execution on a dummy input:
+
+```python
+import torch
+print('\nLoading tacotron2 model from torch.hub.')
+hub_model = torch.hub.load(github='nvidia/DeepLearningExamples', model='nvidia_tacotron2')
+hub_model = hub_model.cuda()
```
[the remark this answers is not in the document]
Comment (soumith) [reply nvidia_deeplearningexamples_tacotron2.md]: considering that it's NVIDIA providing the model, i think it's okay to let this stand.

File: nvidia_deeplearningexamples_tacotron2.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+hub_model.eval()
+inp = torch.randint(low=0, high=148, size=(1,140), dtype=torch.long)
+inp = torch.autograd.Variable(inp).cuda().long()
```
[the remark this answers is not in the document]
Comment (soumith) [reply nvidia_deeplearningexamples_tacotron2.md]: yes, please remove `torch.autograd.Variable(`

File: nvidia_deeplearningexamples_tacotron2.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+Here's a sample execution on a dummy input:
+
+```python
+import torch
+print('\nLoading tacotron2 model from torch.hub.')
+hub_model = torch.hub.load(github='nvidia/DeepLearningExamples', model='nvidia_tacotron2')
+hub_model = hub_model.cuda()
+hub_model.eval()
+inp = torch.randint(low=0, high=148, size=(1,140), dtype=torch.long)
```
Comment (soumith) [inline nvidia_deeplearningexamples_tacotron2.md]: describe what this input is supposed to be, for example see https://pytorch.org/hub/facebookresearch_pytorch-gan-zoo_dcgan/

File: nvidia_deeplearningexamples_waveglow.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+featured_image_1: no-image
+featured_image_2: no-image
+---
+
+```python
+import torch
+hub_model = torch.hub.load(github='nvidia/DeepLearningExamples', model='nvidia_waveglow')
+```
+will load the WaveGlow model pre-trained on [LJ Speech dataset](https://keithito.com/LJ-Speech-Dataset/)
+
```
Comment (soumith) [inline nvidia_deeplearningexamples_waveglow.md]: same thoughts as above. Add details on what the model takes as input and gives as output.
For example see: https://pytorch.org/hub/facebookresearch_pytorch-gan-zoo_dcgan/

File: nvidia_deeplearningexamples_waveglow.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+hub_model = hub_model.cuda()
+hub_model = hub_model.remove_weightnorm(hub_model)
+hub_model.eval()
+inp = torch.randn([1,80,300], dtype=torch.float32).cuda()
+with torch.no_grad():
+    out = hub_model.infer(inp)
+print('\nWaveglow model test output:')
+print(out.size())
+```
+
```
Comment (soumith) [inline nvidia_deeplearningexamples_waveglow.md]: add a visualization of the test output, or some other way to interpret the output

File: scripts/run_pytorch.sh (hunk 1 above)
Comment (soumith) [reply scripts/run_pytorch.sh]: considering this is NVIDIA, i think it's okay

Comment (nv-kkudrynski) [reply scripts/run_pytorch.sh]: thx. I had to revive the above condition in run_pytorch.sh to pass your CI

File: nvidia_deeplearningexamples_tacotron2.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (nv-kkudrynski) [reply nvidia_deeplearningexamples_tacotron2.md]: sure. thx. done

File: nvidia_deeplearningexamples_tacotron2.md (hunk 2 above)
Comment (nv-kkudrynski) [reply nvidia_deeplearningexamples_tacotron2.md]: ok

**Comment by nv-kkudrynski** [comment]
> 1. can we remove NCF from this PR for now -- I dont think the model as it stands is useful for Hub (we are curating), because it isn't really useful for building research on top of it -- or useful to visualize anything about the output

Indeed, NCF is quite specific. We also had discussions about it. It could however be useful if loaded pretrained=False (I added example in readme and provided reference for details about training). Then pre-trained model could serve as reference. Let me know what you think.

> 2. Can you add code that visualizes the outputs of WaveGlow / TacoTron-2. They produce spectrograms, that's great -- it'd be useful to see that. You can use the GAN model that I cleaned up as an example: https://pytorch.org/hub/facebookresearch_pytorch-gan-zoo_dcgan/ . This way, people pulling your model know what to do with it further.

In our flow formatting input/output is not part of the model object, so we need a few internal changes to adapt to your request. We are working on this atm.

**Comment by soumith** [comment]
@nv-kkudrynski i think NCF can serve as a reference with `pretrained=False` if it's easy to train upon. Is the code snippet that shows -- taking a csv dataset or Pandas Dataframe, training the NCF model with it -- going to be about ~20 lines or less? If so, I think it can be worth it for the reader who comes to the hub page. If the reference model is only useful if the data is preprocessed and run through a high-performance set of instructions that are given in the NCF repo, then it's almost better to just point users to the NCF repo directly without a card.

**Comment by soumith** [comment]
even an "easy-finetuning" starting from the pretrained model possibility has potential, but from what I can tell with the dataset, the pre-trained model doesn't have much to gain wrt fine-tuning -- the features aren't generic to generalize to a different set of users + attributes to do recommendation on

**Comment by soumith** [comment]
> In our flow formatting input/output is not part of the model object, so we need a few internal changes to adapt to your request. We are working on this atm.

Great, thanks for the update.
Just FYI, if the input our output is not part of the model object, that is okay. This is why hub entrypoints are callables and not models.
For example, you can have an entry-point for pre-processing, and an entry-point for post-processing.
In the [BERT](https://pytorch.org/hub/huggingface_pytorch-pretrained-bert_bert/) example, they have an entry `bertTokenizer` which is not a model, it's just a pre-processing tokenizer -- but it's available as an entrypoint.
The example showcases taking an input, tokenizing it, and then passing in the output of `bertTokenizer` into `bertModel`. Maybe something like this will help with your case as well, just a thought.

**Comment by nv-kkudrynski** [comment]
> Just FYI, if the input our output is not part of the model object, that is okay. This is why hub entrypoints are callables and not models.

I think there was a discussion about it which concluded with exactly the opposite :) 
Which is reflected in your guidelines at: https://pytorch.org/docs/stable/hub.html:
>Entrypoint function should ALWAYS return a model(nn.module).

But don't worry, update is coming ;)

**Comment by soumith** [comment]
@ailzhang that line needs to be removed I think?

**Comment by nv-kkudrynski** [comment]
I removed NCF from this PR to enable tacotron2/waveglow hopefully before the weekend.
Now the 2 models play very nicely together in the examples - producing sound from input text.

File: nvidia_deeplearningexamples_tacotron2.md (hunk 7 above)
Comment (nv-kkudrynski) [reply nvidia_deeplearningexamples_tacotron2.md]: changed strategy of this example to go from plain text to sound using the two torch.hub models in series

File: nvidia_deeplearningexamples_waveglow.md (hunk 8 above)
Comment (nv-kkudrynski) [reply nvidia_deeplearningexamples_waveglow.md]: changed strategy of this example to go from plain text to sound using the two torch.hub models in series

File: nvidia_deeplearningexamples_waveglow.md (hunk 9 above)
Comment (nv-kkudrynski) [reply nvidia_deeplearningexamples_waveglow.md]: changed strategy of this example to go from plain text to sound using the two torch.hub models in series

**Comment by gottbrath** [comment]
@soumith  - have the changes here addressed all your concerns?

**Comment by soumith** [comment]
thanks so much @nv-kkudrynski , the updated stuff looks good. I might make some additional changes so that it plays well with opening the notebooks in Google Colab.

Additionally, if you are interested as a follow-up (totally up to you and if it makes sense), add end-points for any intermediate features that might make sense for folks to use these models as a feature-extractor / fine-tuner.

**Comment by nv-kkudrynski** [comment]
Thanks to everyone for all the feedback and smooth cooperation!

**Comment by soumith** [comment]
fyi, I have updated the hub card to play well with google colab -- now you can generate the speech and listen to it immediately, all in the colab notebook.

**Comment by nv-kkudrynski** [comment]
Thanks a lot for that!

**Pull request merged**